In [19]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns 

from warnings import filterwarnings
filterwarnings('ignore')

In [20]:
df = pd.read_csv('general_data.csv')
df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      4410 non-null   int64  
 1   Attrition                4410 non-null   object 
 2   BusinessTravel           4410 non-null   object 
 3   Department               4410 non-null   object 
 4   DistanceFromHome         4410 non-null   int64  
 5   Education                4410 non-null   int64  
 6   EducationField           4410 non-null   object 
 7   EmployeeCount            4410 non-null   int64  
 8   EmployeeID               4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

In [22]:
df.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeID                  0
Gender                      0
JobLevel                    0
JobRole                     0
MaritalStatus               0
MonthlyIncome               0
NumCompaniesWorked         19
Over18                      0
PercentSalaryHike           0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           9
TrainingTimesLastYear       0
YearsAtCompany              0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64

In [23]:
df.describe().T

count          mean           std      min  \
Age                      4410.0     36.923810      9.133301     18.0   
DistanceFromHome         4410.0      9.192517      8.105026      1.0   
Education                4410.0      2.912925      1.023933      1.0   
EmployeeCount            4410.0      1.000000      0.000000      1.0   
EmployeeID               4410.0   2205.500000   1273.201673      1.0   
JobLevel                 4410.0      2.063946      1.106689      1.0   
MonthlyIncome            4410.0  65029.312925  47068.888559  10090.0   
NumCompaniesWorked       4391.0      2.694830      2.498887      0.0   
PercentSalaryHike        4410.0     15.209524      3.659108     11.0   
StandardHours            4410.0      8.000000      0.000000      8.0   
StockOptionLevel         4410.0      0.793878      0.851883      0.0   
TotalWorkingYears        4401.0     11.279936      7.782222      0.0   
TrainingTimesLastYear    4410.0      2.799320      1.288978      0.0   
YearsAtCompany           4410.0      7.008163      6.125135      0.0   
YearsSinceLastPromotion  4410.0      2.187755      3.221699      0.0   
YearsWithCurrManager     4410.0      4.123129      3.567327      0.0   

                              25%      50%       75%       max  
Age                         30.00     36.0     43.00      60.0  
DistanceFromHome             2.00      7.0     14.00      29.0  
Education                    2.00      3.0      4.00       5.0  
EmployeeCount                1.00      1.0      1.00       1.0  
EmployeeID                1103.25   2205.5   3307.75    4410.0  
JobLevel                     1.00      2.0      3.00       5.0  
MonthlyIncome            29110.00  49190.0  83800.00  199990.0  
NumCompaniesWorked           1.00      2.0      4.00       9.0  
PercentSalaryHike           12.00     14.0     18.00      25.0  
StandardHours                8.00      8.0      8.00       8.0  
StockOptionLevel             0.00      1.0      1.00       3.0  
TotalWorkingYears            6.00     10.0     15.00      40.0  
TrainingTimesLastYear        2.00      3.0      3.00       6.0  
YearsAtCompany               3.00      5.0      9.00      40.0  
YearsSinceLastPromotion      0.00      1.0      3.00      15.0  
YearsWithCurrManager         2.00      3.0      7.00      17.0

In [24]:
#### Filling Null Values by mean and median 


In [25]:
df['TotalWorkingYears'] = df['TotalWorkingYears'].fillna(11.28) 
# 11.28 is the mean of TotalWorkingYears column 

df['NumCompaniesWorked'] = df['NumCompaniesWorked'].fillna(2.69)
# 2.69 is the median of NumCompaniesWorked

In [26]:
df.isnull().sum()

Age                        0
Attrition                  0
BusinessTravel             0
Department                 0
DistanceFromHome           0
Education                  0
EducationField             0
EmployeeCount              0
EmployeeID                 0
Gender                     0
JobLevel                   0
JobRole                    0
MaritalStatus              0
MonthlyIncome              0
NumCompaniesWorked         0
Over18                     0
PercentSalaryHike          0
StandardHours              0
StockOptionLevel           0
TotalWorkingYears          0
TrainingTimesLastYear      0
YearsAtCompany             0
YearsSinceLastPromotion    0
YearsWithCurrManager       0
dtype: int64

In [27]:
### Dropping unnecessary columns

In [28]:
df = df.drop(['EmployeeCount','EmployeeID', 'Over18','StandardHours'], axis=1)
df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  Gender  JobLevel                    JobRole  \
0          2  Life Sciences  Female         1  Healthcare Representative   
1          1  Life Sciences  Female         1         Research Scientist   
2          4          Other    Male         4            Sales Executive   
3          5  Life Sciences    Male         3            Human Resources   
4          1        Medical    Male         1            Sales Executive   

  MaritalStatus  MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  \
0       Married         131160                 1.0                 11   
1        Single          41890                 0.0                 23   
2       Married         193280                 1.0                 15   
3       Married          83210                 3.0                 11   
4        Single          23420                 4.0                 12   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0                1.0                      6               1   
1                 1                6.0                      3               5   
2                 3                5.0                      2               5   
3                 3               13.0                      5               8   
4                 2                9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4

In [29]:
### Converting Text Column to Categorical Form


In [32]:
from sklearn.preprocessing import LabelEncoder
def encode(cols):
    enc = LabelEncoder()
    for col in cols:
        df[col] = enc.fit_transform(df[col])
encode_cols = ['Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus']
encode(encode_cols)

In [33]:
df.head()

Age  Attrition  BusinessTravel  Department  DistanceFromHome  Education  \
0   51          0               2           2                 6          2   
1   31          1               1           1                10          1   
2   32          0               1           1                17          4   
3   38          0               0           1                 2          5   
4   32          0               2           1                10          1   

   EducationField  Gender  JobLevel  JobRole  MaritalStatus  MonthlyIncome  \
0               1       0         1        0              1         131160   
1               1       0         1        6              2          41890   
2               4       1         4        7              1         193280   
3               1       1         3        1              1          83210   
4               3       1         1        7              2          23420   

   NumCompaniesWorked  PercentSalaryHike  StockOptionLevel  TotalWorkingYears  \
0                 1.0                 11                 0                1.0   
1                 0.0                 23                 1                6.0   
2                 1.0                 15                 3                5.0   
3                 3.0                 11                 3               13.0   
4                 4.0                 12                 2                9.0   

   TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
0                      6               1                        0   
1                      3               5                        1   
2                      2               5                        0   
3                      5               8                        7   
4                      2               6                        0   

   YearsWithCurrManager  
0                     0  
1                     4  
2                     3  
3                     5  
4                     4

In [34]:
## Spliting the Dataset

In [35]:
y = df.iloc[:,1:2]
y.head()

Attrition
0          0
1          1
2          0
3          0
4          0

In [36]:
x = df.drop(['Attrition'], axis=1)
x.head()

Age  BusinessTravel  Department  DistanceFromHome  Education  \
0   51               2           2                 6          2   
1   31               1           1                10          1   
2   32               1           1                17          4   
3   38               0           1                 2          5   
4   32               2           1                10          1   

   EducationField  Gender  JobLevel  JobRole  MaritalStatus  MonthlyIncome  \
0               1       0         1        0              1         131160   
1               1       0         1        6              2          41890   
2               4       1         4        7              1         193280   
3               1       1         3        1              1          83210   
4               3       1         1        7              2          23420   

   NumCompaniesWorked  PercentSalaryHike  StockOptionLevel  TotalWorkingYears  \
0                 1.0                 11                 0                1.0   
1                 0.0                 23                 1                6.0   
2                 1.0                 15                 3                5.0   
3                 3.0                 11                 3               13.0   
4                 4.0                 12                 2                9.0   

   TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
0                      6               1                        0   
1                      3               5                        1   
2                      2               5                        0   
3                      5               8                        7   
4                      2               6                        0   

   YearsWithCurrManager  
0                     0  
1                     4  
2                     3  
3                     5  
4                     4

In [37]:
# Applying Logistic Regression
* Target Variable is Attrition which is Binary Categorical ( 0 or 1 )

SyntaxError: invalid syntax (<ipython-input-37-d90ee6a2e1b4>, line 2)

In [39]:
import statsmodels.api as sm
X1 = sm.add_constant(x)
logistic_model1 = sm.Logit(y,X1)

In [40]:
result1 = logistic_model1.fit()

Optimization terminated successfully.
         Current function value: 0.393016
         Iterations 7


In [41]:
result1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Attrition   No. Observations:                 4410
Model:                          Logit   Df Residuals:                     4390
Method:                           MLE   Df Model:                           19
Date:                Sun, 09 Aug 2020   Pseudo R-squ.:                  0.1102
Time:                        09:04:22   Log-Likelihood:                -1733.2
converged:                       True   LL-Null:                       -1947.9
Covariance Type:            nonrobust   LLR p-value:                 3.382e-79
===========================================================================================
                              coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------
const                       0.0759      0.414      0.183      0.854      -0.735       0.887
Age                        -0.0309      0.007     -4.524      0.000      -0.044      -0.018
BusinessTravel             -0.0177      0.065     -0.270      0.787      -0.146       0.111
Department                 -0.2423      0.081     -2.981      0.003      -0.402      -0.083
DistanceFromHome           -0.0013      0.005     -0.247      0.805      -0.012       0.009
Education                  -0.0628      0.043     -1.473      0.141      -0.146       0.021
EducationField             -0.0966      0.033     -2.894      0.004      -0.162      -0.031
Gender                      0.0860      0.090      0.961      0.337      -0.089       0.262
JobLevel                   -0.0241      0.040     -0.608      0.543      -0.102       0.054
JobRole                     0.0377      0.018      2.108      0.035       0.003       0.073
MaritalStatus               0.5887      0.063      9.323      0.000       0.465       0.712
MonthlyIncome           -1.874e-06   9.56e-07     -1.961      0.050   -3.75e-06   -6.89e-10
NumCompaniesWorked          0.1163      0.018      6.338      0.000       0.080       0.152
PercentSalaryHike           0.0118      0.012      1.005      0.315      -0.011       0.035
StockOptionLevel           -0.0639      0.052     -1.238      0.216      -0.165       0.037
TotalWorkingYears          -0.0574      0.012     -4.819      0.000      -0.081      -0.034
TrainingTimesLastYear      -0.1467      0.035     -4.171      0.000      -0.216      -0.078
YearsAtCompany              0.0120      0.018      0.656      0.512      -0.024       0.048
YearsSinceLastPromotion     0.1322      0.020      6.459      0.000       0.092       0.172
YearsWithCurrManager       -0.1395      0.022     -6.308      0.000      -0.183      -0.096
===========================================================================================
"""

### Inference from Logistic Regression

<hr>

* BusinessTravel, DistanceFromHome, Education, Gender, JobLevel, PercentSalaryHike, StockOptionLevel, YearsAtCompany have **p value more than 0.05** ie, They are Not Significantly Important 
* Age, Department, EducationField, JobRole, MaritalStatus, MonthlyIncome, NumCompaniesWorked, TotalWorkingYears, TrainingTimesLastYear, YearsSinceLastPromotion, YearsWithCurrManager has **p value less than 0.05** ie, They are Significantly Important.
* Out of All the Attributes which are having p values less than 0.05, Age, MaritalStatus, NumCompaniesWorked, TotalWorkingYears, TrainingTimesLastYear, YearsSinceLastPromotion and YearsWithCurrManager are most significant.
* The regression equation is -

***Attrition = 0.075 + (-0.03)Age + (-0.02)BusinessTravel + (-0.24)Department + (-0.0013)DistanceFromHome + (-0.06)Education+ (-0.097)EducationField + (0.086)Gender + (-0.02)JobLevel + (0.038)JobRole + (0.589)MaritalStatus + (-1.8e-06)MonthlyIncome + (0.116)NumCompaniesWorked + (0.012)PercentSalaryHike + (-0.06)StockOptionLevel + (-0.057)TotalWorkingYears + (-0.147)TrainingTimesLastYear + (0.012)YearsAtCompany + (0.132)YearsSinceLastPromotion + (-0.1395)YearsWithCurrManager*** 